File: cpp/client/ads_client.cpp
#include "ads_client.h"
#include "../common/logging.h"
#include <iostream>
#include <thread>
#include <chrono>
#include <random>
#include <map>

static logging::Logger logger("CLIENT");

AdsClient::AdsClient(std::shared_ptr<Channel> channel)
    : stub_(AdsService::NewStub(channel)) {
}

AdsList AdsClient::getAds(const std::string& query, const std::string& asin_id, const std::string& understanding) {
    logging::Timer overall_timer("bidirectional_stream");
    ClientContext context;
    std::unique_ptr<ClientReaderWriter<Context, AdsList>> stream(stub_->GetAds(&context));
    
    std::string start_message = logging::LogContext()
        .add("query", query)
        .add("asin_id", asin_id)
        .add("understanding_provided", !understanding.empty())
        .build("Opening bidirectional stream");
    logger.info(start_message);
    
    // Send Context messages in a separate thread
    std::thread sender([this, &stream, &query, &asin_id, &understanding, &overall_timer]() {
        sendContextMessages(stream.get(), query, asin_id, understanding, overall_timer);
    });
    
    // Receive AdsList messages with timeout logic
    AdsList result = receiveAdsListWithTimeout(stream.get(), overall_timer);
    
    // Wait for sender thread to complete
    sender.join();
    
    // Finish the call
    Status status = stream->Finish();
    if (!status.ok()) {
        std::string error_message = logging::LogContext()
            .add("error_code", status.error_code())
            .add("error_message", status.error_message())
            .add("elapsed_ms", overall_timer.elapsed_ms())
            .build("GetAds RPC failed");
        logger.error(error_message);
    } else {
        std::string success_message = logging::LogContext()
            .add("total_duration_ms", overall_timer.elapsed_ms())
            .build("GetAds RPC completed successfully");
        logger.info(success_message);
    }
    
    return result;
}

void AdsClient::sendContextMessages(ClientReaderWriter<Context, AdsList>* stream,
                                  const std::string& query, 
                                  const std::string& asin_id, 
                                  const std::string& understanding,
                                  const logging::Timer& overall_timer) {
    // Send first Context message
    Context context1;
    context1.set_query(query);
    context1.set_asin_id(asin_id);
    context1.set_understanding(""); // Empty initially
    
    std::string first_context_message = logging::LogContext()
        .add("context_number", 1)
        .add("understanding_empty", true)
        .add("elapsed_ms", overall_timer.elapsed_ms())
        .build("Sending Context message");
    logger.info(first_context_message);
    
    if (stream->Write(context1)) {
        logger.debug("First Context message sent successfully");
    } else {
        logger.error("Failed to send first Context message");
        return;
    }
    
    // Wait 50ms before sending second message
    logger.debug("Waiting 50ms before second Context message");
    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    
    // Send second Context message with understanding
    Context context2;
    context2.set_query(query);
    context2.set_asin_id(asin_id);
    context2.set_understanding(understanding);
    
    std::string second_context_message = logging::LogContext()
        .add("context_number", 2)
        .add("understanding_length", understanding.length())
        .add("elapsed_ms", overall_timer.elapsed_ms())
        .build("Sending Context message");
    logger.info(second_context_message);
    
    if (stream->Write(context2)) {
        logger.debug("Second Context message sent successfully");
    } else {
        logger.error("Failed to send second Context message");
        return;
    }
    
    // Half-close the stream (client side done sending)
    stream->WritesDone();
    
    std::string half_close_message = logging::LogContext()
        .add("elapsed_ms", overall_timer.elapsed_ms())
        .build("Half-closed client stream");
    logger.info(half_close_message);
}

AdsList AdsClient::receiveAdsListWithTimeout(ClientReaderWriter<Context, AdsList>* stream, 
                                            const logging::Timer& overall_timer) {
    std::map<uint32_t, AdsList> adsListBuffer; // Buffer by version number
    AdsList currentAdsList;
    
    // Generate random timeout (30-120ms jittered)
    int timeoutMs = generateRandomTimeout();
    std::string timeout_message = logging::LogContext()
        .add("timeout_ms", timeoutMs)
        .add("min_timeout", 30)
        .add("max_timeout", 120)
        .build("Generated random timeout for result selection");
    logger.info(timeout_message);
    
    auto startTime = std::chrono::steady_clock::now();
    auto timeoutDuration = std::chrono::milliseconds(timeoutMs);
    
    // Read AdsList messages until timeout
    while (true) {
        auto currentTime = std::chrono::steady_clock::now();
        auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(currentTime - startTime);
        
        if (elapsed >= timeoutDuration) {
            std::string timeout_reached_message = logging::LogContext()
                .add("timeout_ms", timeoutMs)
                .add("elapsed_ms", overall_timer.elapsed_ms())
                .add("versions_received", adsListBuffer.size())
                .build("Timeout reached, proceeding with available results");
            logger.info(timeout_reached_message);
            break;
        }
        
        AdsList adsList;
        if (stream->Read(&adsList)) {
            uint32_t version = adsList.version();
            bool is_replacement = adsListBuffer.find(version) != adsListBuffer.end();
            
            std::string receive_message = logging::LogContext()
                .add("version", version)
                .add("ads_count", adsList.ads_size())
                .add("elapsed_ms", overall_timer.elapsed_ms())
                .add("is_replacement", is_replacement)
                .build("Received AdsList");
            logger.info(receive_message);
            
            // Log debug details about the ads if debug level is enabled
            if (logger.is_debug_enabled()) {
                for (int i = 0; i < adsList.ads_size(); i++) {
                    const auto& ad = adsList.ads(i);
                    std::string ad_message = logging::LogContext()
                        .add("version", version)
                        .add("ad_index", i)
                        .add("asin_id", ad.asin_id())
                        .add("ad_id", ad.ad_id())
                        .add("score", ad.score())
                        .build("Ad details");
                    logger.debug(ad_message);
                }
            }
            
            // Buffer the AdsList by version (replace older versions)
            if (is_replacement) {
                std::string replace_message = logging::LogContext()
                    .add("version", version)
                    .add("old_ads_count", adsListBuffer[version].ads_size())
                    .add("new_ads_count", adsList.ads_size())
                    .build("Replaced AdsList in buffer");
                logger.debug(replace_message);
            }
            
            adsListBuffer[version] = adsList;
            currentAdsList = adsList; // Keep track of the most recent
        } else {
            // Stream ended, break out of loop
            std::string stream_end_message = logging::LogContext()
                .add("elapsed_ms", overall_timer.elapsed_ms())
                .add("versions_received", adsListBuffer.size())
                .build("Stream ended");
            logger.info(stream_end_message);
            break;
        }
    }
    
    // Log buffer state for debugging
    std::ostringstream versions_oss;
    for (const auto& pair : adsListBuffer) {
        if (versions_oss.tellp() > 0) versions_oss << ",";
        versions_oss << pair.first;
    }
    
    std::string buffer_state_message = logging::LogContext()
        .add("buffer_size", adsListBuffer.size())
        .add("available_versions", versions_oss.str())
        .add("elapsed_ms", overall_timer.elapsed_ms())
        .build("Buffer state at timeout");
    logger.debug(buffer_state_message);
    
    // Return the latest version available
    if (!adsListBuffer.empty()) {
        auto latestEntry = adsListBuffer.rbegin(); // Get highest version
        uint32_t finalVersion = latestEntry->first;
        AdsList finalResult = latestEntry->second;
        
        std::string final_result_message = logging::LogContext()
            .add("selected_version", finalVersion)
            .add("ads_count", finalResult.ads_size())
            .add("total_duration_ms", overall_timer.elapsed_ms())
            .add("versions_considered", adsListBuffer.size())
            .build("FINAL RESULT: Selected AdsList");
        logger.info(final_result_message);
        
        // Log performance summary
        std::string perf_message = logging::LogContext()
            .add("operation", "bidirectional_stream")
            .add("total_duration_ms", overall_timer.elapsed_ms())
            .add("timeout_used_ms", timeoutMs)
            .add("versions_received", adsListBuffer.size())
            .add("final_version", finalVersion)
            .build("Performance summary");
        logger.info(perf_message);
        
        return finalResult;
    } else {
        std::string no_result_message = logging::LogContext()
            .add("total_duration_ms", overall_timer.elapsed_ms())
            .add("timeout_ms", timeoutMs)
            .add("buffer_size", adsListBuffer.size())
            .build("FINAL RESULT: No AdsList received within timeout");
        logger.warn(no_result_message);
        return AdsList(); // Return empty AdsList
    }
}

int AdsClient::generateRandomTimeout() {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> dis(30, 120);
    return dis(gen);
}

void AdsClient::shutdown() {
    // Nothing specific to clean up for this implementation
    logger.info("Client shutdown completed");
}

File: java/src/main/java/com/example/ads/client/AdsClient.java
package com.example.ads.client;

import ads.Ads;
import ads.AdsServiceGrpc;
import com.example.ads.common.LoggingConfig;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ThreadLocalRandom;
import java.util.logging.Logger;
import java.util.logging.Level;

/**
 * gRPC bidirectional streaming client for the ads service.
 * Implements the client-side logic for progressive ad refinement.
 */
public class AdsClient {
    private static final Logger logger = LoggingConfig.configureLogger(AdsClient.class, "CLIENT");
    
    private final ManagedChannel channel;
    private final AdsServiceGrpc.AdsServiceStub asyncStub;
    
    // Configuration constants
    private static final int CONTEXT_DELAY_MS = 50;
    private static final int MIN_TIMEOUT_MS = 30;
    private static final int MAX_TIMEOUT_MS = 120;
    
    /**
     * Construct client connecting to the ads server at {@code host:port}.
     */
    public AdsClient(String host, int port) {
        this(ManagedChannelBuilder.forAddress(host, port)
                .usePlaintext()
                .build());
    }
    
    /**
     * Construct client for accessing ads server using the existing channel.
     */
    AdsClient(ManagedChannel channel) {
        this.channel = channel;
        this.asyncStub = AdsServiceGrpc.newStub(channel);
    }
    
    /**
     * Get ads using bidirectional streaming with progressive context refinement.
     * 
     * @param query The search query
     * @param asinId The product identifier
     * @param understanding The refined understanding (will be sent in second context)
     * @return The final AdsList or null if no ads received
     */
    public Ads.AdsList getAds(String query, String asinId, String understanding) {
        LoggingConfig.Timer overallTimer = new LoggingConfig.Timer("bidirectional_stream");
        
        String startMessage = new LoggingConfig.LogContext()
            .add("query", query)
            .add("asin_id", asinId)
            .add("understanding_provided", !understanding.isEmpty())
            .build("Starting bidirectional stream");
        logger.info(startMessage);
        
        // Thread-safe map to store received AdsList by version
        final ConcurrentHashMap<Integer, Ads.AdsList> adsBuffer = new ConcurrentHashMap<>();
        final CountDownLatch finishedLatch = new CountDownLatch(1);
        
        // Response observer to handle incoming AdsList messages
        StreamObserver<Ads.AdsList> responseObserver = new StreamObserver<Ads.AdsList>() {
            @Override
            public void onNext(Ads.AdsList adsList) {
                long elapsedMs = overallTimer.elapsedMs();
                int version = (int) adsList.getVersion();
                boolean isReplacement = adsBuffer.containsKey(version);
                
                String receiveMessage = new LoggingConfig.LogContext()
                    .add("version", version)
                    .add("ads_count", adsList.getAdsCount())
                    .add("elapsed_ms", elapsedMs)
                    .add("is_replacement", isReplacement)
                    .build("Received AdsList");
                logger.info(receiveMessage);
                
                // Log debug details about the ads if debug level is enabled
                if (logger.isLoggable(Level.FINE)) {
                    for (int i = 0; i < adsList.getAdsCount(); i++) {
                        Ads.Ad ad = adsList.getAds(i);
                        String adMessage = new LoggingConfig.LogContext()
                            .add("version", version)
                            .add("ad_index", i)
                            .add("asin_id", ad.getAsinId())
                            .add("ad_id", ad.getAdId())
                            .add("score", String.format("%.3f", ad.getScore()))
                            .build("Ad details");
                        logger.fine(adMessage);
                    }
                }
                
                // Buffer the AdsList, replacing any previous version
                Ads.AdsList previousVersion = adsBuffer.put(version, adsList);
                if (previousVersion != null) {
                    String replaceMessage = new LoggingConfig.LogContext()
                        .add("version", version)
                        .add("old_ads_count", previousVersion.getAdsCount())
                        .add("new_ads_count", adsList.getAdsCount())
                        .build("Replaced AdsList in buffer");
                    logger.fine(replaceMessage);
                }
            }
            
            @Override
            public void onError(Throwable t) {
                String errorMessage = new LoggingConfig.LogContext()
                    .add("error_type", t.getClass().getSimpleName())
                    .add("error_message", t.getMessage())
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .build("GetAds RPC failed");
                logger.log(Level.WARNING, errorMessage, t);
                finishedLatch.countDown();
            }
            
            @Override
            public void onCompleted() {
                String completeMessage = new LoggingConfig.LogContext()
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .add("versions_received", adsBuffer.size())
                    .build("Server completed response stream");
                logger.info(completeMessage);
                finishedLatch.countDown();
            }
        };
        
        // Start the bidirectional stream
        StreamObserver<Ads.Context> requestObserver = asyncStub.getAds(responseObserver);
        
        try {
            // Send first Context message with empty understanding
            LoggingConfig.Timer contextTimer = new LoggingConfig.Timer("first_context_send");
            Ads.Context firstContext = Ads.Context.newBuilder()
                    .setQuery(query)
                    .setAsinId(asinId)
                    .setUnderstanding("")
                    .build();
            
            String firstContextMessage = new LoggingConfig.LogContext()
                    .add("context_number", 1)
                    .add("understanding_empty", true)
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .build("Sending Context message");
            logger.info(firstContextMessage);
            requestObserver.onNext(firstContext);
            
            // Wait 50ms before sending second context
            logger.fine("Waiting " + CONTEXT_DELAY_MS + "ms before second Context message");
            Thread.sleep(CONTEXT_DELAY_MS);
            
            // Send second Context message with filled understanding
            Ads.Context secondContext = Ads.Context.newBuilder()
                    .setQuery(query)
                    .setAsinId(asinId)
                    .setUnderstanding(understanding)
                    .build();
            
            String secondContextMessage = new LoggingConfig.LogContext()
                    .add("context_number", 2)
                    .add("understanding_length", understanding.length())
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .build("Sending Context message");
            logger.info(secondContextMessage);
            requestObserver.onNext(secondContext);
            
            // Half-close the client side of the stream
            requestObserver.onCompleted();
            String halfCloseMessage = new LoggingConfig.LogContext()
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .build("Half-closed client stream");
            logger.info(halfCloseMessage);
            
            // Generate random timeout between 30-120ms
            int timeoutMs = ThreadLocalRandom.current().nextInt(MIN_TIMEOUT_MS, MAX_TIMEOUT_MS + 1);
            String timeoutMessage = new LoggingConfig.LogContext()
                    .add("timeout_ms", timeoutMs)
                    .add("min_timeout", MIN_TIMEOUT_MS)
                    .add("max_timeout", MAX_TIMEOUT_MS)
                    .build("Generated random timeout for result selection");
            logger.info(timeoutMessage);
            
            // Wait for the random timeout
            Thread.sleep(timeoutMs);
            
            // Return the most recent AdsList (highest version number)
            Ads.AdsList finalResult = null;
            int maxVersion = 0;
            
            // Log buffer state for debugging
            String bufferStateMessage = new LoggingConfig.LogContext()
                    .add("buffer_size", adsBuffer.size())
                    .add("available_versions", adsBuffer.keySet().toString())
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .build("Buffer state at timeout");
            logger.fine(bufferStateMessage);
            
            for (Ads.AdsList adsList : adsBuffer.values()) {
                if (adsList.getVersion() > maxVersion) {
                    maxVersion = (int) adsList.getVersion();
                    finalResult = adsList;
                }
            }
            
            if (finalResult != null) {
                String finalResultMessage = new LoggingConfig.LogContext()
                        .add("selected_version", finalResult.getVersion())
                        .add("ads_count", finalResult.getAdsCount())
                        .add("total_duration_ms", overallTimer.elapsedMs())
                        .add("versions_considered", adsBuffer.size())
                        .build("FINAL RESULT: Selected AdsList");
                logger.info(finalResultMessage);
                
                // Log performance summary
                String perfMessage = new LoggingConfig.LogContext()
                        .add("operation", "bidirectional_stream")
                        .add("total_duration_ms", overallTimer.elapsedMs())
                        .add("timeout_used_ms", timeoutMs)
                        .add("versions_received", adsBuffer.size())
                        .add("final_version", finalResult.getVersion())
                        .build("Performance summary");
                logger.info(perfMessage);
            } else {
                String noResultMessage = new LoggingConfig.LogContext()
                        .add("total_duration_ms", overallTimer.elapsedMs())
                        .add("timeout_ms", timeoutMs)
                        .add("buffer_size", adsBuffer.size())
                        .build("FINAL RESULT: No AdsList received within timeout");
                logger.warning(noResultMessage);
            }
            
            return finalResult;
            
        } catch (InterruptedException e) {
            String interruptMessage = new LoggingConfig.LogContext()
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .add("buffer_size", adsBuffer.size())
                    .build("Client interrupted");
            logger.log(Level.WARNING, interruptMessage, e);
            Thread.currentThread().interrupt();
            return null;
        } catch (RuntimeException e) {
            String runtimeErrorMessage = new LoggingConfig.LogContext()
                    .add("error_type", e.getClass().getSimpleName())
                    .add("elapsed_ms", overallTimer.elapsedMs())
                    .build("Runtime error during stream processing");
            logger.log(Level.SEVERE, runtimeErrorMessage, e);
            // Cancel RPC
            requestObserver.onError(e);
            throw e;
        }
    }
    
    /**
     * Shutdown the client channel.
     */
    public void shutdown() throws InterruptedException {
        channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
    }
    
    /**
     * Main method to run the client.
     */
    public static void main(String[] args) throws Exception {
        String host = "localhost";
        int port = 8080;
        
        // Parse command line arguments if provided
        if (args.length >= 2) {
            host = args[0];
            port = Integer.parseInt(args[1]);
        }
        
        AdsClient client = new AdsClient(host, port);
        try {
            // Example usage
            String query = "coffee maker";
            String asinId = "B000123456";
            String understanding = "user wants high-quality coffee brewing equipment";
            
            Ads.AdsList result = client.getAds(query, asinId, understanding);
            
            if (result != null) {
                System.out.println(String.format("Final result: AdsList version %d with %d ads", 
                    result.getVersion(), result.getAdsCount()));
                
                for (Ads.Ad ad : result.getAdsList()) {
                    System.out.println(String.format("  Ad: asinId=%s, adId=%s, score=%.3f", 
                        ad.getAsinId(), ad.getAdId(), ad.getScore()));
                }
            } else {
                System.out.println("No ads received");
            }
            
        } finally {
            client.shutdown();
        }
    }
}